src/charm.py

```diff
@@ -152,26 +179,63 @@ def _on_hydra_pebble_ready(self, event: WorkloadEvent) -> None:
 
         self.unit.status = MaintenanceStatus("Configuring resources")
 
-        self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
-        logger.info("Pebble plan updated with new configuration, replanning")
-
-        try:
-            self._container.replan()
-        except ChangeError as err:
-            logger.error(str(err))
-            self.unit.status = BlockedStatus("Failed to replan, please consult the logs")
+        current_layer = self._container.get_plan()
+        new_layer = self._hydra_layer
+        if current_layer.services != new_layer.services:
+            self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            try:
+                self._container.replan()
+            except ChangeError as err:
+                logger.error(str(err))
+                self.unit.status = BlockedStatus("Failed to replan, please consult the logs")
+                return
+
+        if not self._hydra_service_is_created:
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Hydra service")
+            logger.info("Hydra service is absent. Deferring the event.")
             return
 
-        if self.database.is_resource_created():
-            self._container.push(self._hydra_config_path, self._render_conf_file(), make_dirs=True)
-            self._container.start(self._container_name)
-            self.unit.status = ActiveStatus()
+        if not self.model.relations[self._db_relation_name]:
+            self.unit.status = BlockedStatus("Missing required relation with postgresql")
             return
 
-        if self.model.relations[self._db_relation_name]:
+        if not self.database.is_resource_created():
             self.unit.status = WaitingStatus("Waiting for database creation")
-        else:
-            self.unit.status = BlockedStatus("Missing required relation with postgresql")
+            event.defer()
+            return
+
+        self._container.push(self._hydra_config_path, self._render_conf_file(), make_dirs=True)
+        self._container.restart(self._container_name)
+        self.unit.status = ActiveStatus()
+
+    def _update_hydra_endpoints_relation_data(self, event: RelationEvent) -> None:
```

Comment:
Not sure if we actually care about the types(probably not?), but the correct type for the event here is `HookEvent`

Reply:
Are you sure? This method can be called by ingress or hydra endpoints relation events, which are all of `RelationEvent` class.
`RelationEvent` is based on `HookEvent` though